File: Stacks/simplify_path.py
class SimplifyPath:
    def simplify_path(self, path: str) -> str:
        segments = path.split("/")
        stack = []
        for segment in segments:
            if segment == "" or segment == ".":
                continue
            elif segment == "..":
                if stack:
                    stack.pop()
            else:
                stack.append(segment)

        return "/" + "/".join(stack)


# Time complexity: O(n) where n is the length of the path string
# Space complexity: O(n) in the worst case if all segments are valid directory names

path = "/.../a/../b/c/../d/./"
print(SimplifyPath().simplify_path(path))

File: Stacks/SimplifyPath.java
package Stacks;

import java.util.ArrayDeque;
import java.util.Deque;

public class SimplifyPath {
  public String simplifyPath(String path) {
    Deque<String> stack = new ArrayDeque<>();
    String[] segments = path.split("/");
    for (String segment : segments) {
      if (segment.isEmpty() || segment.equals(".")) {
        continue;
      } else if (segment.equals("..")) {
        if (!stack.isEmpty())
          stack.removeLast();
      } else {
        stack.add(segment);
      }
    }

    StringBuilder simplifiedPath = new StringBuilder();
    for (String segment : stack) {
      simplifiedPath.append("/");
      simplifiedPath.append(segment);
    }

    return simplifiedPath.length() > 0 ? simplifiedPath.toString() : "/";
  }
}

// Time complexity O(n) where n is the length of the input path
// Space complexity O(n) in the worst case when all segments are added to the
// stack